pkg/provider/provider.go

```diff
@@ -1,221 +1,393 @@
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_ROLE", nil),
+			},
+			"validate_default_parameters": {
+				Type:        schema.TypeBool,
+				Description: "ValidateDefaultParameters disable the validation checks for Database, Schema, Warehouse and Role at the time a connection is established. Can also be sourced from the `SNOWFLAKE_VALIDATE_DEFAULT_PARAMETERS` environment variable.",
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_VALIDATE_DEFAULT_PARAMETERS", false),
+			},
+			"params": {
+				Type:        schema.TypeMap,
+				Description: "Sets other connection (i.e. session) parameters. [Parameters](https://docs.snowflake.com/en/sql-reference/parameters)",
+				Optional:    true,
+			},
+			"client_ip": {
+				Type:        schema.TypeString,
+				Description: "IP address for network checks. Can also be sourced from the `SNOWFLAKE_CLIENT_IP` environment variable.",
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_CLIENT_IP", nil),
+			},
+			"protocol": {
+				Type:        schema.TypeString,
+				Description: "Either http or https, defaults to https. Can also be sourced from the `SNOWFLAKE_PROTOCOL` environment variable.",
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_PROTOCOL", nil),
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					switch strings.ToLower(val.(string)) {
+					case "http", "https":
+						return nil, nil
+					default:
+						errs := append(errs, fmt.Errorf("%q must be one of http or https", key))
+						return warns, errs
+					}
+				},
+			},
+			"host": {
+				Type:        schema.TypeString,
+				Description: "Supports passing in a custom host value to the snowflake go driver for use with privatelink. Can also be sourced from the `SNOWFLAKE_HOST` environment variable. ",
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_HOST", nil),
+			},
+			"port": {
+				Type:        schema.TypeInt,
+				Description: "Support custom port values to snowflake go driver for use with privatelink. Can also be sourced from the `SNOWFLAKE_PORT` environment variable. ",
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_PORT", nil),
+			},
+			"authenticator": {
+				Type:        schema.TypeString,
+				Description: "Specifies the [authentication type](https://pkg.go.dev/github.com/snowflakedb/gosnowflake#AuthType) to use when connecting to Snowflake. Valid values include: Snowflake, OAuth, ExternalBrowser, Okta, JWT, TokenAccessor, UsernamePasswordMFA. Can also be sourced from the `SNOWFLAKE_AUTHENTICATOR` environment variable.",
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_AUTHENTICATOR", nil),
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
```

Comment:
the same here, I think all such functions should be extracted and not defined in place, I also think that this may be a topic for a followup task, which could be added as a subtask of SNOW-933289

Reply:
i suppose we could, firstly there aren't too many validators, and secondly, the framework does validators a bit differently. so even if we extract this code and put it somewhere else, it would eventually become obsolete.